### PR TITLE
feat: remove Python 3.7 support

### PR DESCRIPTION
**Pull request by rwgk**
## Description
Remove Python 3.7 support entirely. ([Official end-of-life](https://devguide.python.org/versions/): 2023-06-27)

NOTE: We're losing testing Clang versions 3, 7, 9, 10, and GCC version 7, but we're deliberately deferring reviewing the C++ code to remove related `#ifdef`s.

See also (below): https://github.com/pybind/pybind11/pull/5191#issuecomment-2187330041
________

For easy future reference, these were the `git grep` commands used to pin-point lines that may need to be changed:

```bash
#! /bin/bash
set -x
git grep 0x0307
git grep 0x0308
git grep PY_MINOR_VERSION
git grep PYPY_VERSION
git grep -I '[^0-9]37[^0-9]'
git grep -I '[^0-9]38[^0-9]'
git grep -I -E '3\.7'
git grep -I -E '3\.8'
git grep -I -E '\(3, 7'
git grep -I -E '\(3, 8'
git grep -I -E '3[^A-Za-z0-9.]+7'
git grep -I -E '3[^A-Za-z0-9.]+8'
```

## Suggested changelog entry:

```rst
Support for Python 3.7 was removed. (Official end-of-life: 2023-06-27)
```

### CODE REVIEW

**Comment by rwgk** [comment]
Overview of all compiler versions used before this PR, and with this PR @ commit 2f24de56d447712a4d294b57152a14bd6b48be49:

* Clang release dates: https://releases.llvm.org/

* GNU release dates: https://gcc.gnu.org/releases.html

```
 AppleClang 15.0.0.15000040
-Clang 3.6.2    16 Jul 2015
-Clang 3.7.1     5 Jan 2016
-Clang 3.9.1    23 Dec 2016
 Clang 5.0.2
-Clang 7.1.0    10 May 2019
-Clang 9.0.1    20 Dec 2019
-Clang 10.0.1    6 Aug 2020
 Clang 11.1.0
 Clang 12.0.1
 Clang 13.0.1
 Clang 14.0.6
 Clang 15.0.7
 Clang 16.0.6
 Clang 18.1.6 with GNU-like command-line
 Clang 18.1.7
 Clang 19.0.0
-GNU 7.5.0      November 14, 2019
-GNU 8.3.0      February 22, 2019
 GNU 8.5.0
 GNU 9.4.0
+GNU 10.2.1
 GNU 10.5.0
 GNU 11.4.0
 GNU 11.4.1
 GNU 12.4.0
 GNU 13.2.0
 GNU 13.2.1
 GNU 13.3.0
 Intel 2021.10.0.20230609
 MSVC 19.29.30154.0
 MSVC 19.40.33811.0
 NVHPC 23.5.0
```

**Comment by rwgk** [comment]
Tracking simple findings:

* ubuntu-18.04 `apt-get update && apt-get install -y gcc` produces `g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0`
 
* ubuntu-20.04 `apt-get update && apt-get install -y clang` produces `clang version 10.0.0-4ubuntu1`

**Comment by henryiii** [comment]
I think this next release might be a good time to remove C++11 support and increase our minimum supported compilers.

**Comment by rwgk** [comment]
> I think this next release might be a good time to remove C++11 support and increase our minimum supported compilers.

Awesome. That's what I was thinking, too. I'll work on this PR more after the 2.13 release has baked in a bit.

**Comment by henryiii** [comment]
I think we can also bump CMake support. Personally, I've been going 3.15+ everywhere as that is what was required to build scikit-build-core. Though another option is 3.12+ (FindPython). I think we has someone request 3.10 or 3.11 before, might be worth checking.

If we did go 3.15+, we maybe could start to phase out classic FindPythonInterp support.

**Comment by rwgk** [comment]
> If we did go 3.15+, we maybe could start to phase out classic FindPythonInterp support.

I just looked up: CMake 3.15.0 publication date was 17 July 2019

Does that look correct? — That's already 5 years ago (almost at least), and before Python 3.8 was released (Oct 2019). — I think it's much more likely to be helpful than harmful to make CMake 3.15 the minimum, and shed cruft.

**Comment by henryiii** [comment]
Here's the support table for CMake, with darker red for older than 3.15:

cmake versions

**Comment by rwgk** [comment]
My thinking:

* I want to wait about one week, JIC we need more 2.13 patch releases.
* Then I'll make a fresh pass through this PR, looking for 3.7 and 3.8 version numbers of all sorts once more.
* Then merge this PR, documenting in the PR description that we're losing testing with Clang 3.6 through 10, and GCC 7 as a consequence of dropping Python 3.7 support, but we're deliberately deferring cleaning up ifdefs for another release.
* Then have a follow-on PR to bump the cmake minimum to 3.15 (independently from dropping Python 3.7) and clean out cmake cruft.

**Comment by henryiii** [comment]
We’ve already made a 2.13 branch, master is 2.14. No rush, but waiting isn’t required. Does save ci time though if the branches are functionally identical.

No need to defer cleanup, we have plenty of time for this to sit in master. We should specify and test minimum versions where possible. I’m hoping we get a MSVC bump. :)

But that can also be a follow-up.

File: pyproject.toml [hunk 1]
```diff
@@ -30,7 +30,7 @@ ignore_missing_imports = true
 
 
 [tool.pylint]
-master.py-version = "3.7"
+master.py-version = "3.8"
```
Comment (Skylion007) [inline pyproject.toml:33]: ruff has a Python version too, right? otherwise UP036 would complain above

Comment (Skylion007) [reply pyproject.toml:33]: Yeah, it does: `target-version = "py37"`

Comment (henryiii) [reply pyproject.toml:33]: This will go away if we move to scikit-build-core after 0.10 comes out. ;)

Comment (rwgk) [reply pyproject.toml:33]: Thanks for catching this @Skylion007!

Fixed with commit 49580cfa977eaef89b95d070e59214ebc9a05c43

That made me realize we need to add something like this to the list of useful `git grep` commands, for the next time when we want to drop a Python version:

```
git grep -I '[^0-9]37[^0-9]'
git grep -I '[^0-9]38[^0-9]'
```

Comment (Skylion007) [reply pyproject.toml:33]: It will also go away if define the minimum python version in pyproject.toml standard location instead of setup.cfg :)

**Review by Skylion007 — APPROVED** [review]

**Comment by rwgk** [comment]
Thanks @Skylion007 !

**Pull request merged**